roadmap/implementers-guide/src/pvf-prechecking.md

```diff
@@ -0,0 +1,50 @@
+
+This has the following implications:
+- In backing, inclusion may be slow due to backing groups being unable to execute the block
+- In approval checking, there may be many no-shows, leading to slow finality
+- In disputes, neither side may reach supermajority. Nobody will get slashed and the chain will not be reverted or finalized.
+
+As a result of this issue we need a fairly hard guarantee that the PVFs of registered parachains/threads can be compiled within a reasonable amount of time.
+
+## Solution
+
+The problem is solved by having a pre-checking process which is run when a new validation code is included in the chain. A new PVF can be added in two cases:
+
+- A new parachain or parathread is registered.
+- An existing parachain or parathread signalled an upgrade of its validation code.
+
+Before any of those processes finish, the PVF pre-checking vote is initiated. The PVF pre-checking vote is identified by the PVF code hash that is being voted on. If there is already PVF pre-checking process running, then no
+new PVF pre-checking vote will be started. Instead, the process just subscribes to the existing vote.
+
+The pre-checking vote can be concluded either by obtaining a supermajority or if it expires.
+
+Each validator checks the list of PVFs available for voting. The vote is binary, i.e. accept or reject a given PVF. As soon as the supermajority of votes are collected for one of the sides of the vote, the voting is concluded in that direction and the effects of the voting are enacted.
+
+Only validators from the active set can participate in the vote. The set of active validators can change each session. That's why we reset the votes each session. A voting that observed a certain number of sessions will be rejected.
+
+The effects of the PVF accepting depend on the operations requested it:
+
+1. All onboardings subscribed to the approved PVF pre-checking process will get scheduled and after passing 2 session boundaries they will be onboarded.
+1. All upgrades subscribed to the approved PVF pre-checking process will get scheduled very similarly to the existing process. Upgrades with pre-checking are really the same process that is just delayed by the time required for pre-checking voting. In case of instant approval the mechanism is exactly the same.
+
+In case, PVF pre-checking process was concluded with rejection, then all the requesting operations get cancelled. For onboarding it means it gets without movement: the lifecycle of such parachain is terminated on the `Onboarding` state and after rejection the lifecycle is none. That in turn means that the caller can attempt registering the parachain once more. For upgrading it means that the upgrade process is aborted: that flashes go-ahead signal with `Abort` flag. Rejection leads to removing the allegedly bad validation code from the chain storage.
```

Comment:
This paragraph is hard to understand at least for me. I don't really get what `For onboarding it means it gets without movement:` really means.